static/src/javascripts/projects/common/modules/crosswords/crossword.js

```diff
@@ -0,0 +1,778 @@
+            } else {
+                newDirection = 'down';
+            }
+            this.focusClue(x, y, newDirection);
+        }
+    }
+
+    onCheat(): void {
+        this.allHighlightedClues().forEach(clue => this.cheat(clue));
+        this.save();
+    }
+
+    onCheck(): void {
+        // 'Check this' checks single and grouped clues
+        this.allHighlightedClues().forEach(clue => this.check(clue));
+        this.save();
+    }
+
+    onSolution(): void {
+        this.props.data.entries.forEach(clue => this.cheat(clue));
+        this.save();
+    }
+
+    onCheckAll(): void {
+        this.props.data.entries.forEach(clue => this.check(clue));
+        this.save();
+    }
+
+    onClearAll(): void {
+        this.setState({
+            grid: helpers.mapGrid(this.state.grid, cell => {
+                cell.value = '';
+                return cell;
+            }),
+        });
+
+        this.save();
+    }
+
+    onClearSingle(): void {
+        // Merge arrays of cells from all highlighted clues
+        // const cellsInFocus = _.flatten(_.map(this.allHighlightedClues(), helpers.cellsForEntry, this));
+        const cellsInFocus = helpers.getClearableCellsForClue(
+            this.state.grid,
+            this.clueMap,
+            this.props.data.entries,
+            this.clueInFocus()
+        );
+
+        this.setState({
+            grid: helpers.mapGrid(this.state.grid, (cell, gridX, gridY) => {
+                if (cellsInFocus.some(c => c.x === gridX && c.y === gridY)) {
+                    cell.value = '';
+                }
+                return cell;
+            }),
+        });
+
+        this.save();
+    }
+
+    onToggleAnagramHelper(): void {
+        // only show anagram helper if a clue is active
+        if (!this.state.showAnagramHelper) {
+            if (this.clueInFocus()) {
+                this.setState({
+                    showAnagramHelper: true,
+                });
+            }
+        } else {
+            this.setState({
+                showAnagramHelper: false,
+            });
+        }
+    }
+
+    onClickHiddenInput(event: Event): void {
+        const focussed = this.state.cellInFocus;
+
+        this.onSelect(focussed.x, focussed.y);
+
+        /* We need to handle touch seperately as touching an input on iPhone does not fire the
+         click event - listen for a touchStart and preventDefault to avoid calling onSelect twice on
+         devices that fire click AND touch events. The click event doesn't fire only when the input is already focused */
+        if (event.type === 'touchstart') {
+            event.preventDefault();
+        }
+    }
+
+    setGridHeight(): void {
+        if (!this.$gridWrapper) {
+            this.$gridWrapper = $(React.findDOMNode(this.refs.gridWrapper));
+        }
+
+        if (
+            isBreakpoint({
+                max: 'tablet',
+            })
+        ) {
+            fastdom.read(() => {
```

Comment:
Is this `fastdom.read` call necessary?

Reply:
I have no idea. How would I go about finding out?

Reply:
I don't think there's any real way, it's a judgement thing... I probably should have phrased this as "This `fastdom.read` call is not necessary 😅  

Since this is a big conversion, maybe we should leave it for now (way to go, talking myself out of this recommending this change!)

Reply:
@SiAdcock Once we've done the fastdom upgrade, we could look into [strict mode](https://github.com/wilsonpage/fastdom#strict-mode) for the dev environment and start eliminating useless/ de-optimized calls (writes in the read queue and the other way around). Feels like there are many of them.

I agree: in this case `fastdom.read` seems not to be necessary in this case.